src/renderer/App.tsx

```diff
@@ -120,6 +123,52 @@ function Sidebar() {
 }
 
 function Topbar() {
+  const wallet = useWallet();
+  const validator = useAppSelector(selectValidatorNetworkState);
+  const { net } = validator;
+
+  useEffect(() => {
+    const airdropIfNeeded = async () => {
+      if (validator.status !== NetStatus.Running) return;
+      let amount = 0;
+
+      // arbitrary -- want to let people top up
+      // on devnet SOL when they can
+      let airdropThreshold = 50;
+      switch (net) {
+        case Net.Dev:
+          // limit per devnet aidrop request
+          amount = 2;
+          break;
+        case Net.Test:
+          // limit per testnet airdrop request
+          amount = 1;
+          // arbitrary
+          airdropThreshold = 5;
+          break;
+        case Net.MainnetBeta:
+          return;
+        default:
+          amount = 1000;
+      }
+      const solConn = new sol.Connection(netToURL(net));
+      if (wallet.publicKey) {
+        try {
+          const balance = await solConn.getBalance(wallet.publicKey);
+          if (balance < airdropThreshold) {
+            solConn.requestAirdrop(
+              wallet.publicKey,
+              sol.LAMPORTS_PER_SOL * amount
+            );
+          }
+        } catch (e) {
+          logger.error(e);
+        }
+      }
+    };
+    airdropIfNeeded();
+  }, [net, wallet.publicKey, validator]);
```

Comment:
So this will happen once per change of network connection / wallet? (and so if at that point the network was offline, we won't try again?

Reply:
Yea. And validator status. So if a brand new local validator starts up without airdrop, it will airdrop again. Was quick to implement and felt a bit less aggressive than polling.

Reply:
polling bad :)